### PR TITLE
[BIT-484] Add TextCausalLMNext

### PR DESCRIPTION
**Pull request by opentaco**
Adds the TextCausalLMNext Synapse, as well as the tokenizer_utils needed to encode/decode the synapse. Both core_server and core_validator are running successfully on Nobunaga with this branch.

TextCausalLMNext sends a user-defined topk of server token phrases for input context continuation, which validator/client uses for validated generation where server phrase probabilities are evaluated against a groundtruth continuation. Supports backward gradients from validator/client to server.

Check detailed commit explanations.

### CODE REVIEW

File: bittensor/_dendrite/dendrite_impl.py [hunk 1]
```diff
@@ -436,7 +436,80 @@ def text_causal_lm (
         self.update_stats( formatted_endpoints, synapses, formatted_inputs, outputs, codes, times )
         return outputs[0], codes[0], times[0]
 
-    def text_last_hidden_state( 
+    def text_causal_lm_next(
```
Comment (Eugene-hu) [inline bittensor/_dendrite/dendrite_impl.py:439]: Should we expect this synapse to be called on its own or always in conjunction with CausalLM?

File: bittensor/_synapse/__init__.py [hunk 2]
```diff
@@ -101,6 +102,38 @@ def TextCausalLM (
...
+        topk: int = 4096,
+        forward_request_serializer_type: 'bittensor.proto.Serializer.Type' = bittensor.proto.Serializer.MSGPACK,
+        forward_response_serializer_type: 'bittensor.proto.Serializer.Type' = bittensor.proto.Serializer.MSGPACK,
+        backward_request_serializer_type: 'bittensor.proto.Serializer.Type' = bittensor.proto.Serializer.MSGPACK,
+        backward_response_serializer_type: 'bittensor.proto.Serializer.Type' = bittensor.proto.Serializer.MSGPACK,
+    ) -> TextCausalLMNext:
+        """ Factory function which returns a TextCausalLMNext synapse adapter given arguments.
+            Args:
+                topk (:obj:`int`):
+                    Specifies the number of topk server token phrases to return.
```
Comment (Eugene-hu) [inline bittensor/_synapse/__init__.py:116]: A little bit confused about the topk variable. Are they the topk logits of the server's tokenizer?

File: bittensor/_synapse/text_causallmnext_impl.py [hunk 3]
```diff
@@ -0,0 +1,170 @@
...
+        ):
+            raise ValueError(f"backward_request_gradient.shape must be in "
+                             f"[>={forward_request_tensor.shape[0]} x (2 x {self.topk} + 1)], "
+                             f"got: {backward_request_gradient.size(0)} for synapse: {self}")
+
+    def encode_forward_request_tensor(self, forward_request_tensor: torch.Tensor) -> torch.Tensor:
+        return forward_request_tensor
+
+    def decode_forward_request_tensor(self, forward_request_tensor: torch.Tensor) -> torch.Tensor:
+        return forward_request_tensor
```
Comment (Eugene-hu) [inline bittensor/_synapse/text_causallmnext_impl.py:140]: Do we need a encoding or decoding function?

File: bittensor/_synapse/text_causallmnext_impl.py [hunk 4]
```diff
@@ -0,0 +1,170 @@
...
+                len(forward_request_tensor.shape) != 2 or
+                forward_request_tensor.shape[0] == 0 or
+                forward_request_tensor.shape[1] == 0
+        ):
+            raise ValueError(f"forward_request_tensor.shape must be in [-1, -1], "
+                             f"got: {list(forward_request_tensor.shape)} for synapse: {self}")
+
+    def check_forward_response_tensor(self, forward_request_tensor, forward_response_tensor):
+        # forward_request_tensor: [batch_size, sequence_len]
+        # forward_response_tensor: [ >= batch_size * (2 * topk + 1)]
```
Comment (Eugene-hu) [inline bittensor/_synapse/text_causallmnext_impl.py:113]: The response tensor is a 1-D tensor?

**Review by Eugene-hu — APPROVED** [review]
Approved, left you a few questions

File: bittensor/_dendrite/dendrite_impl.py (hunk 1 above)
Comment (opentaco) [reply bittensor/_dendrite/dendrite_impl.py:439]: TextCausalLMNext can be called independently of TextCausalLM, but for initial tests they'll both run so comparisons can be performed.

File: bittensor/_synapse/__init__.py (hunk 2 above)
Comment (opentaco) [reply bittensor/_synapse/__init__.py:116]: Yes, topk logits/phrases, under the assumption that each logit represents a text phrase (the server represents this with a single token with associated probability, but it could be multiple std tokens (a token phrase) at the validator).

File: bittensor/_synapse/text_causallmnext_impl.py (hunk 3 above)
Comment (opentaco) [reply bittensor/_synapse/text_causallmnext_impl.py:140]: Currently, encoding and decoding is best performed at the origin, that is server and validator forward() functions respectively. The reason being that variable length phrases need to otherwise be converted to padded tensors, simply for the purpose of passing it to the synapse encode_forward_request_tensor(), where it would need to be decoded and then re-encoded and compacted, thus duplicating work best done immediately at the origin.

File: bittensor/_synapse/text_causallmnext_impl.py (hunk 4 above)
Comment (opentaco) [reply bittensor/_synapse/text_causallmnext_impl.py:113]: Yes, the response is 1-D to efficiently compact variable length token phrases by omitting padding that would otherwise be required to make it n-D.
